### PR TITLE
Issue 27 attach textual stack trace to snapshots

### PR DESCRIPTION
**Pull request by bocasti**
Fix proposal for issue #27

Added stackTrace property to Snapshot class.
=> Initialize stackTrace as exception/context gets converted as snapshot
=> store stack stract as an additinal piece of bytes to be read after the fuel context
=> added unit test
=> updated UI in order to provide an additional "Stack trace" button
=> Removed duplication of specific mime types

### CODE REVIEW

**Review by noha — COMMENTED** [review]
Why do we need to access the stack trace separate and have the need to introduce another header and such? Would include the stack trace in the JSON be a problem?

File: source/SnapDump-Core/SDMemorySnapshot.class.st [hunk 1]
```diff
@@ -48,13 +48,15 @@ SDMemorySnapshot >> serializeContext: aContext on: stream [
 SDMemorySnapshot >> setContext: aContext exceptionClassName: exceptionClassName [
 	| method |
 	context := aContext.
+	self
+		stackTrace:
+			(String streamContents: [ :stream | context errorReportOn: stream ]).
```
Comment (noha) [inline source/SnapDump-Core/SDMemorySnapshot.class.st:53]: I wonder if we should not restrict the stack trace and fuel to include only a certain amount of stack information.

**Comment by bocasti** [comment]
> Why do we need to access the stack trace separate and have the need to introduce another header and such? Would include the stack trace in the JSON be a problem?

Good question, I was not sure about the way to go.
The advantage of having it separate is that if this stack trace gets bigger at one point, we are not forced to get it by default.
We can stream it on demand.
That's why I mimicked the way we were doing with the context

However currently this stack trace is a simple character string that is limited in size.
So technically I do not see obstacle in having it part of the basic JSON properties

File: source/SnapDump-Core/SDMemorySnapshot.class.st (hunk 1 above)
Comment (bocasti) [reply source/SnapDump-Core/SDMemorySnapshot.class.st:53]: What do you think should not be part of the stack or the context ?

**Comment by noha** [comment]
The get stack trace test is failing in pharo8. We should make it compatible with pharo8

**Comment by noha** [comment]
I'm not sure if it is better to put it into JSON. So let's go this way for now. I've added #30 to make this at least solid

File: source/SnapDump-Core/SDMemorySnapshot.class.st (hunk 1 above)
Comment (noha) [reply source/SnapDump-Core/SDMemorySnapshot.class.st:53]: I'm not sure. I just had often the case that the stack got too big to report although I'm interested only in the first 50 stack frames or such. But I have nothing concrete in mind. And the textual stack trace grows slower than the fuel dump so let's think  about it for a while

**Comment by bocasti** [comment]
> I'm not sure if it is better to put it into JSON. So let's go this way for now. I've added #30 to make this at least solid

Ok, right !

File: source/SnapDump-Client/SDHTTPStore.class.st [hunk 2]
```diff
@@ -116,11 +116,20 @@ SDHTTPStore >> shouldReportSnapshot: snapshot [
...
 
+{ #category : #accessing }
+SDHTTPStore >> snapshotStackFor: aSnapshot [ 
+
+	^ (self client
+		headerAt: 'Accept' put: SDStore mimeTypeStack;
+		url: (self urlForSnapshot: aSnapshot);
+		get) utf8Decoded
+]
+
```
Comment (stonesong) [inline source/SnapDump-Client/SDHTTPStore.class.st:132]: At first I found confusing that there are two methods on the client but only one endpoint on the server checking on the mimeType. It's ok but I find it confusing that the client has to use "utf8Decoded" on the response. Why isn't thqt done on server side ?

**Review by stonesong — APPROVED** [review]

File: source/SnapDump-Client/SDHTTPStore.class.st (hunk 2 above)
Comment (noha) [reply source/SnapDump-Client/SDHTTPStore.class.st:132]: The server uses a streaming entity which means it does not read to whole file in memory before responding but streams it. And it sets a binary mime-type. It is possible to make it different but as it is done with the client classes I don't see a big problem here

**Review by noha — APPROVED** [review]

**Pull request merged**